pkg/cache/sql/db/client.go

```diff
@@ -0,0 +1,321 @@
+	Exec(query string, args ...any) (sql.Result, error)
+	Prepare(query string) (*sql.Stmt, error)
+	Close() error
+}
+
+// Closable Closes an underlying connection and returns an error on failure
+type Closable interface {
+	Close() error
+}
+
+// Rows represents sql rows. It exposes method to navigate the rows, read their outputs, and close them.
+type Rows interface {
+	Next() bool
+	Err() error
+	Close() error
+	Scan(dest ...any) error
+}
+
+// TXClient represents a sql transaction. The TXClient must manage rollbacks as rollback functionality is not exposed.
+type TXClient interface {
+	StmtExec(stmt transaction.Stmt, args ...any) error
+	Exec(stmt string, args ...any) error
+	Commit() error
+	Stmt(stmt *sql.Stmt) transaction.Stmt
+	Cancel() error
+}
+
+// Encryptor encrypts a slice of bytes. It is expected to use data-encryption-key (DEK) which it used to encrypt the slice.
+// It should use a key-encryption-key (KEK) to then encrypt the DEK. The encrypted data, nonce used to encrypt the data,
+// encrypted dek, nonce used to decrypt the dek should be returned, in that order. On failure an error should by returned.
+type Encryptor interface {
+	Encrypt([]byte) ([]byte, []byte, []byte, []byte, error)
+}
+
+// Decryptor decrypts encrypted data given: some encrypted data, nonce used to encrypt the data, an encrypted
+// data-encryption-key (DEK), nonce used to encrypt the DEK. The KEK is not provided. Decryptor is expected to use its
+// receiver to decrypt the DEK using its KEK.
+type Decryptor interface {
+	Decrypt([]byte, []byte, []byte, []byte) ([]byte, error)
+}
+
+var backoffRetry = wait.Backoff{Duration: 50 * time.Millisecond, Factor: 2, Steps: 10}
+
+// NewClient returns a Client. If the given connection is nil then a default one will be created.
+func NewClient(c Connection, encryptor Encryptor, decrypter Decryptor) (*Client, error) {
+	client := &Client{
+		encryptor: encryptor,
+		decryptor: decrypter,
+	}
+	if c != nil {
+		client.conn = c
+		return client, nil
+	}
+	err := client.NewConnection()
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
+// Prepare prepares the given string into a sql statement on the client's connection.
+func (c *Client) Prepare(stmt string) *sql.Stmt {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	prepared, err := c.conn.Prepare(stmt)
+	if err != nil {
+		panic(errors.Errorf("Error preparing statement: %s\n%v", stmt, err))
+	}
+	return prepared
+}
+
+// QueryForRows queries the given stmt with the given params and returns the resulting rows. The query wil be retried
+// given a sqlite busy error.
+func (c *Client) QueryForRows(ctx context.Context, stmt transaction.Stmt, params ...any) (*sql.Rows, error) {
+	c.connLock.RLock()
+	defer c.connLock.RUnlock()
+	var rows *sql.Rows
+	var err error
+
+	err = wait.ExponentialBackoff(backoffRetry, func() (bool, error) {
+		rows, err = stmt.QueryContext(ctx, params...)
+		if err != nil {
+			sqlErr, ok := err.(*sqlite.Error)
```

Comment:
I'm not sure this works because the type of `sqlite.Error` is:
```
type Error struct {
	msg  string
	code int
}
``` 
so it's not a wrapped `error`.

Reply:
It implements error and is returned the sqlite driver returns them.

Reply:
yup